doc/neps/ufunc-overrides.rst

```diff
@@ -213,83 +370,128 @@ They have indirect interactions, however, because NumPy's
 most numerical classes, the easiest way to override binary operations is
 thus to define ``__array_ufunc__`` and override the corresponding
 Ufunc. The class can then, like :class:`ndarray` itself, define the
-binary operators in terms of Ufuncs. Here, one has to take some care.
-E.g., the simplest implementation would be::
+binary operators in terms of Ufuncs. Here, one has to take some care to
+ensure that one allows for other classes to indicate they are not
+compatible, i.e., implementations should be something like::
 
     class ArrayLike(object):
+        ...
         def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
             ...
             return result
-        ...
-        def __mul__(self, other):
-            return self.__array_ufunc__(np.multiply, '__call__', self, other)
-
-Suppose now, however, that ``other`` is class that does not know how to
-deal with arrays and ufuncs, but does know how to do multiplication::
 
-    class MyObject(object):
-        __array_ufunc__ = None
+        # Option 1: call ufunc directly
```

Comment:
> As an aside, the ArrayLike example in the gist uses an unconditional nested ufunc dispatch --- which puts it in the type hierarchy above all other types, which probably is not intended. 

Indeed, this is not what I actually want. Two different classes that implement this exact same interface would have an ill-defined result when they are added together.

I was hoping for some sort of cooperative behavior without setting an explicit type hierarchy, but I realize now that that is impossible. If `ArrayLike` wants to be able to wrap arbitrary other array types, it should expose a registration function that allows other array types to say that they can be wrapped by `ArrayLike` instead of trying to wrap everything.